// @infura/sdk/dist/src/services/ipfsService.js
"use strict";
/*!
 * Copyright(c) ConsenSys Software Inc.
 * Copyright(c) https://consensys.net/
 * MIT Licensed
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable */
var fs_1 = __importDefault(require("fs"));
var index_1 = require("../lib/Logger/index");
var utils_1 = require("../lib/utils");
var ipfsServerClient_1 = __importDefault(require("../../e2e/utils/utils.ts/ipfsServerClient"));
var axios_1 = __importDefault(require("axios"));
var IPFS = /** @class */ (function () {
    function IPFS(_a) {
        var projectId = _a.projectId, apiKeySecret = _a.apiKeySecret;
        if (!projectId) {
            index_1.log.throwArgumentError(index_1.Logger.message.no_infura_projectID_supplied, 'projectId', projectId, {
                location: index_1.Logger.location.IPFSSERVICE_CONSTRUCTOR,
            });
        }
        if (!apiKeySecret) {
            index_1.log.throwArgumentError(index_1.Logger.message.no_infura_projectSecret_supplied, 'apiKeySecret', apiKeySecret, {
                location: index_1.Logger.location.IPFSSERVICE_CONSTRUCTOR,
            });
        }
        this.ipfsServerClient = new ipfsServerClient_1.default({
            baseURL: 'https://ipfs.infura.io:5001/api/v0/',
            projectId: projectId || '',
            apiKeySecret: apiKeySecret || '',
        });
    }
    /** Upload free content data on ipfs
     * @param {string} metadata any string
     * @returns {Promise<string>} Ipfs hash of the stored data
     */
    IPFS.prototype.uploadContent = function (_a) {
        var source = _a.source;
        return __awaiter(this, void 0, void 0, function () {
            var _b;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _b = "ipfs://".concat;
                        return [4 /*yield*/, this.ipfsServerClient.add(source)];
                    case 1: return [2 /*return*/, _b.apply("ipfs://", [_c.sent()])];
                }
            });
        });
    };
    /** Upload file on ipfs
     * @param {string} metadata path to local file or url
     * @returns {Promise<string>} Ipfs hash of the stored data
     */
    IPFS.prototype.uploadFile = function (_a) {
        var source = _a.source;
        return __awaiter(this, void 0, void 0, function () {
            var response, _b, inputSrc, _c, error_1;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        _d.trys.push([0, 5, , 6]);
                        if (!(0, utils_1.isURI)(source)) return [3 /*break*/, 3];
                        return [4 /*yield*/, axios_1.default.get(source, { responseType: 'stream' })];
                    case 1:
                        response = _d.sent();
                        _b = "ipfs://".concat;
                        return [4 /*yield*/, this.ipfsServerClient.add(response.data)];
                    case 2: return [2 /*return*/, _b.apply("ipfs://", [_d.sent()])];
                    case 3:
                        if (!fs_1.default.existsSync(source)) {
                            throw index_1.Logger.message.unexisting_file;
                        }
                        inputSrc = fs_1.default.createReadStream(source);
                        _c = "ipfs://".concat;
                        return [4 /*yield*/, this.ipfsServerClient.add(inputSrc)];
                    case 4: return [2 /*return*/, _c.apply("ipfs://", [_d.sent()])];
                    case 5:
                        error_1 = _d.sent();
                        return [2 /*return*/, index_1.log.throwArgumentError(index_1.Logger.message.an_error_occured_with_ipfs_api, 'file', source, {
                                location: index_1.Logger.location.IPFSSERVICE_UPLOADFILE,
                                error: "".concat(error_1),
                            })];
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    /** upload array of metadata on ipfs
     * @param {Array<any>} metadata an array of valid JSON Metadata
     * @returns {Promise<string>} Ipfs hash of the stored data
     */
    IPFS.prototype.uploadArray = function (_a) {
        var sources = _a.sources, _b = _a.isErc1155, isErc1155 = _b === void 0 ? false : _b;
        return __awaiter(this, void 0, void 0, function () {
            var files, _c, error_2;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        if (sources.length <= 0) {
                            return [2 /*return*/, index_1.log.throwArgumentError(index_1.Logger.message.array_should_not_be_empty, 'sources', sources, {
                                    location: index_1.Logger.location.IPFSSERVICE_UPLOADDIRECTORY,
                                })];
                        }
                        _d.label = 1;
                    case 1:
                        _d.trys.push([1, 3, , 4]);
                        files = sources.map(function (source, index) {
                            return {
                                path: isErc1155 ? "".concat(index, ".json") : "".concat(index),
                                content: source,
                            };
                        });
                        _c = "ipfs://".concat;
                        return [4 /*yield*/, this.ipfsServerClient.addAll(files)];
                    case 2: return [2 /*return*/, _c.apply("ipfs://", [_d.sent(), "/"])];
                    case 3:
                        error_2 = _d.sent();
                        return [2 /*return*/, index_1.log.throwError(index_1.Logger.message.an_error_occured_with_ipfs_api, index_1.Logger.code.IPFS, {
                                location: index_1.Logger.location.IPFSSERVICE_UPLOADDIRECTORY,
                                error: error_2,
                            })];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    return IPFS;
}());
exports.default = IPFS;
